AWS Cognito provider

ralphsomeday opened the pull request:
#### Description of what you did:
fix #6822

alexandrebodin commented:
Closing for inactivity @ralphsomeday  Let me know when you can takle this again so I reopen it ;)

JiboStore commented:
Hi, I am new to strapi (just evaluating) and would need to integrate AWS Cognito user-pool as its auth provider (is it possible) but the documentation is minimum.

@ralphsomeday is your PR for cognito integration? Would you please share any resources you use to get it up?

ralphsomeday commented:
@JiboStore i will review the comments from @alexandrebodin and update the PR.

ralphsomeday commented:
@alexandrebodin can you re-open it?

ralphsomeday commented:
![image](https://user-images.githubusercontent.com/61593456/92587395-e25dc100-f297-11ea-9b88-4786685180ea.png)


![image](https://user-images.githubusercontent.com/61593456/92587621-323c8800-f298-11ea-9ba0-1557dcdef9f3.png)

@alexandrebodin how can we solve the merge issue with Travis? thanks

ralphsomeday commented:
@alexandrebodin would you mind reviewing this PR? this will allow to add support for a new provider AWS Cognito.

alexandrebodin reviewed (changes requested):
Hi, thanks for this PR. Using extra config folder is not pactical as it requires more user work. I think this should be set in the plugins.js users-permissions configuration. We could do it a better cleaner buy allowing to overxwritte any provider config from the configuraation directly like so:

`./config/plugins.js`
```js
module.exports = () => ({
  users-permissions: {
    grantProviders: {
      cognito: { // or any other provider
         subdomain: 'azdaz',
       }
    }
  }
})
```

This way we only merge the configs together instead of having to had code for each new properties that will come up in the future

ralphsomeday commented:
@alexandrebodin made it even easier by adding an extra field in the provider configuration and added a translation key in all supported languages see below screenshot:

![image](https://user-images.githubusercontent.com/61593456/93895668-5f555580-fcf0-11ea-8a7e-572add28300f.png)

ralphsomeday commented:
@alexandrebodin any chance this can be reviewed so that we can add support for cognito auth provider in next release?

alexandrebodin commented:
Hi, your PR contains a ton of conflicts right now. Once it is fixed we will review it again. From What I saw already adding the input in the UI won't scale as we can't just add every possible input one PR at a time. I think we should still allow passing those configuration in the configs instead of the UI.

ralphsomeday commented:
@alexandrebodin the conflicts are only one translation key added ("PopUpForm.Providers.subdomain.label": "Host URI (Subdomain)") in the translation files, the auth file has been cleaned up to remove the changes and go back to its original shape and the bootstrap has just added the new AWS Cognito provider. I do not see why this won't scale at UI level, any new provider you are adding is going in the bootstrap and this is a new provider that is now working which allows companies using cognito as an idp to be able to connect it with strapi.

alexandrebodin commented:
I wasn't clear. I meant we won't be able to add input for every single options avaialbe for each providers. In this case I only see in the diff that you added the cognito provider. no issue with that but the PR is just not mergeable so we won't approve it until cleaned up :)

ralphsomeday commented:
@alexandrebodin the thing is that the subdomain is not a custom property of cognito it is provided with grant and can be configured for some providers that needs it. So i guess this is something we can leave in the UI. For the rest let me know what you would like me to have cleaned up so that this PR can go ahead? The subdomain in the UI will allow to configure different providers that may require different subdomains directly in the UI way more handy than having it in the configuration of the plugin.

alexandrebodin commented:
You need to fix all the conflicts so the PR can go ahead

alexandrebodin commented:
Not there yet :/

ralphsomeday commented:
@alexandrebodin should be better now no more merge conflicts. Can you let me know if all is fine from your side after the travis has passed? thanks

ralphsomeday commented:
@alexandrebodin should be good to go

alexandrebodin commented:
@ralphsomeday testing your PR right now. It doesn't have the subdomain input anymore

ralphsomeday commented:
> @ralphsomeday testing your PR right now. It doesn't have the subdomain input anymore

@alexandrebodin you mean you are not seeing the subdomain input field in the AWS Cognito provider?

alexandrebodin commented:
> > @ralphsomeday testing your PR right now. It doesn't have the subdomain input anymore
> 
> @alexandrebodin you mean you are not seeing the subdomain input field in the AWS Cognito provider?

Yep exactly

ralphsomeday commented:
@alexandrebodin i see this on the latest version was testing on 3.0.x let me investigate this and come back to you.

ralphsomeday commented:
@alexandrebodin this should be it. Now working with the latest branch. Let me know if anything is not working. thanks

soupette commented on packages/strapi-plugin-users-permissions/admin/src/containers/Providers/utils/createProvidersArray.js:
```diff
@@ -3,10 +3,14 @@ import { sortBy } from 'lodash';
   return sortBy(
     Object.keys(data).reduce((acc, current) => {
-      const { icon: iconName, enabled } = data[current];
+      const { icon: iconName, enabled, subdomain } = data[current];
```
```
const { icon: iconName, enabled, subdomain } = data[current];
```

soupette commented on packages/strapi-plugin-users-permissions/admin/src/containers/Providers/utils/createProvidersArray.js:
```diff
@@ -3,10 +3,14 @@ import { sortBy } from 'lodash';
       const icon = iconName === 'envelope' ? ['fas', 'envelope'] : ['fab', iconName];
 
-      acc.push({ name: current, icon, enabled });
+      if (subdomain) {
```
The condition here is not needed you can use null for the subdomain

```
acc.push({ name: current, icon, enabled, subdomain });

return acc;
```

soupette replied:
I am sorry you could leave this file as you did it originally.

soupette commented on packages/strapi-plugin-users-permissions/admin/src/containers/Providers/index.js:
```diff
@@ -30,6 +30,7 @@ const ProvidersPage = () => {
   const buttonSubmitRef = useRef(null);
   const [showForm, setShowForm] = useState(false);
   const [providerToEditName, setProviderToEditName] = useState(null);
+  const [providerIsWithSubdomain, setProviderIsWithSubdomain] = useState(false);
```
The useState hook is not needed here you can replace it by the following:


```

import { get, has, upperFirst } from 'lodash';
const isProviderWithSubdomain = useMemo(() => {
    if (!providerToEditName) {
      return false;
    }

    const providerToEdit = providers.find(obj => obj.name === providerToEditName);

    return has(providerToEdit, 'subdomain');
  }, [providers, providerToEditName]);
```


Then you will just need to update the formToRender constant:

```
const formToRender = useMemo(() => {
  if (providerToEditName === 'email') {
    return forms.email;
  }

  if (isProviderWithSubdomain) {
    return forms.providersWithSubdomain;
  }

  return forms.providers;
}, [providerToEditName, isProviderWithSubdomain]);
```

ralphsomeday replied:
for the `return has(...)` I had to add     return has(providerToEdit, 'subdomain') && providerToEdit.subdomain !== undefined; as the previous code in the createProviderArray is adding the subdomain property to all providers.

soupette replied:
I am sorry I have forgotten to remove my comment on the createProvidersArray file if you keep your old code (really sorry about that) the has condition should work.

ralphsomeday replied:
@soupette rolled back to my previous code

soupette commented on packages/strapi-plugin-users-permissions/admin/src/containers/Providers/index.js:
```diff
@@ -97,6 +96,7 @@ const ProvidersPage = () => {
     provider => {
       if (canUpdate) {
         setProviderToEditName(provider.name);
+        setProviderIsWithSubdomain(provider.subdomain && provider.subdomain !== '');
```
With the comment from above you can remove this line.

soupette commented on packages/strapi-plugin-users-permissions/admin/src/containers/Providers/index.js:
```diff
@@ -105,6 +105,7 @@ const ProvidersPage = () => {
 
   const handleClosed = useCallback(() => {
     setProviderToEditName(null);
+    setProviderIsWithSubdomain(false);
```
Same here.

soupette commented on packages/strapi-plugin-users-permissions/admin/src/containers/Providers/utils/forms.js:
```diff
@@ -57,6 +57,17 @@ const forms = {
           required: true,
         },
       },
+      {
```
Instead of adding a new input here I think it's better to add a new key: `providersWithSubdomain` so you have a specific form and a dedicated yup schema:


```
const forms = {
  email: {},
  providers: {},
  providersWithSubdomain: {
    form: [
      {
        autoFocus: true,
        label: getTrad('PopUpForm.Providers.enabled.label'),
        name: 'enabled',
        type: 'bool',
        description: getTrad('PopUpForm.Providers.enabled.description'),
        size: { xs: 6 },
        validations: {
          required: true,
        },
      },
      {
        autoFocus: false,
        label: getTrad('PopUpForm.Providers.key.label'),
        name: 'key',
        type: 'text',
        placeholder: getTrad('PopUpForm.Providers.key.placeholder'),
        size: { xs: 12 },
        validations: {
          required: true,
        },
      },
      {
        autoFocus: false,
        label: getTrad('PopUpForm.Providers.secret.label'),
        name: 'secret',
        type: 'text',
        placeholder: getTrad('PopUpForm.Providers.secret.placeholder'),
        size: { xs: 12 },
        validations: {
          required: true,
        },
      },
      {
        autoFocus: false,
        label: getTrad('PopUpForm.Providers.subdomain.label'),
        name: 'subdomain',
        type: 'text',
        placeholder: getTrad('PopUpForm.Providers.subdomain.placeholder'),
        size: { xs: 12 },
        validations: {
          required: true,
        },
      },
      {
        autoFocus: false,
        label: getTrad('PopUpForm.Providers.redirectURL.front-end.label'),
        placeholder: 'http://www.client-app.com',
        name: 'callback',
        type: 'text',
        size: { xs: 12 },
        validations: {
          required: true,
        },
      },
      {
        label: {
          id: getTrad('PopUpForm.Providers.redirectURL.label'),
          params: {
            provider: 'VK',
          },
        },
        name: 'noName',
        type: 'text',
        validations: {},
        size: {
          xs: 12,
        },
        disabled: true,
      },
    ],
    schema: yup.object().shape({
      enabled: yup.bool().required(translatedErrors.required),
      key: yup.string().when('enabled', {
        is: true,
        then: yup.string().required(translatedErrors.required),
        otherwise: yup.string(),
      }),
      secret: yup.string().when('enabled', {
        is: true,
        then: yup.string().required(translatedErrors.required),
        otherwise: yup.string(),
      }),
      subdomain: yup.string().when('enabled', {
        is: true,
        then: yup.string().required(translatedErrors.required),
        otherwise: yup.string(),
      }),
      callback: yup.string().when('enabled', {
        is: true,
        then: yup.string().required(translatedErrors.required),
        otherwise: yup.string(),
      }),
    }),
  }
}
```

Then it will simplify the index.js file.

soupette commented on packages/strapi-plugin-users-permissions/admin/src/containers/Providers/index.js:
```diff
@@ -219,14 +220,40 @@ const ProvidersPage = () => {
           <form onSubmit={handleSubmit}>
             <Row>
               {formToRender.form.map(input => {
+                if (input.name === 'subdomain') {
```
Regarding the comment I have added in the utils/forms.js  file, this line won't be needed so you can remove them.

soupette reviewed (changes requested):
Thanks a lot for your contribution!

In order to simplify the logic you can simply create a new type of form for the providers that have a subdomain with this you can also update the validation schema without adding more complexity to the code.

Then, in the index.js file you will just need to add a condition in order to map the correct with the correct provider.

Let me know if my comments are not clear enough.

ralphsomeday commented:
@soupette should now be GTM

soupette reviewed (approved):
LGTM for the front-end part!

Thanks a lot for your contribution  👏

alexandrebodin reviewed (changes requested):
Hey, the code looks good to me, 

One last thing though. I think we should add a section in the doc for this provider to explain how to set it up on the aws side (configure the oauth2 conf, authorization flow and such...) cc @derrickmehaffy

derrickmehaffy commented:
> Hey, the code looks good to me,
> 
> One last thing though. I think we should add a section in the doc for this provider to explain how to set it up on the aws side (configure the oauth2 conf, authorization flow and such...) cc @derrickmehaffy

Yup exactly, if there is any non-standard steps (or just in general for all providers merged into the mono-repo) we will need any relevant information added to the docs. Bad link due to how vuepress handles markdown but somewhere around here: https://github.com/strapi/strapi/blob/master/docs/v3.x/plugins/users-permissions.md#providers

ralphsomeday commented:
@alexandrebodin and @derrickmehaffy doc is updated please review and let me know if anything need to be amended or added.

petersg83 commented on docs/v3.x/plugins/users-permissions.md:
```diff
@@ -317,6 +317,42 @@ Wait a few seconds while the application is created.
+- Feel in your callback URL and Sign out URL with the value `http://localhost:1337/connect/cognito/callback` or the one provided by your AWS Cognito provider in Strapi
+- In the **Oauth 2.0** section select `Authorization code grant` and `Implicit grant` for the **Allowed OAuth Flows** and select `email`, `openid` and `profile` for the **Allowed OAuth Scopes**
+- You can now click on **Save changes** and a link to the **Launch Hosted UI** will now be visible. You can click on it in order to display the AWS Cognito login page. Copy the URL of that login page (everything between `https` and `/login`). This will be the **Host URI (Subdomain)** value for your strapi configuration
+
```
The hosted UI was no available because the Domain Name was not set. Can you add a step for that please?

ralphsomeday replied:
Sure

petersg83 commented on docs/v3.x/plugins/users-permissions.md:
```diff
@@ -317,6 +317,42 @@ Wait a few seconds while the application is created.
+
+#### Strapi configuration
+
+- Visit the User Permissions provider settings page <br> [http://localhost:1337/admin/plugins/users-permissions/providers](http://localhost:1337/admin/plugins/users-permissions/providers)
```
The settings have moved to `http://localhost:1337/admin/settings/users-permissions/providers`

petersg83 commented on docs/v3.x/plugins/users-permissions.md:
```diff
@@ -860,3 +896,7 @@ module.exports = {
+
+## Custom providers HOST URI (subdomain) configuration
+
+Some external Identity providers like AWS Cognito will require you to provide the link to the IDP host url (e.g. your_idp_name.auth.eu-central-1.amazoncognito.com) so that you can access the IDP login page using `htts://your_strapi_url/connect/your_custom_provider_name`. In order to do that you will have to fill in that value in the configuration popup of the provider in the Host URI (subdomain)  ...
```
Why did you add that? The doc for aws isn't enough?

ralphsomeday replied:
yes it is, it was added before the doc for AWS was written. Will remove it.

petersg83 reviewed (changes requested):
Hi, 
Thanks for you contribution !
It doesn't work on my side, I've been through the whole process but at the end I have "unknown provider". I think some code is missing in the PR here: https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/services/Providers.js#L452

ralphsomeday commented on docs/v3.x/plugins/users-permissions.md:
```diff
@@ -317,6 +317,42 @@ Wait a few seconds while the application is created.
+- You should now have an **App client id** and by clicking on the button **Show Details** you will be able to see the **App client secret**. Do copy those two values **App client id** and **App client secret** somewhere for later use when configuring the AWS Cognito provider in Strapi.
+- Go to the **App integration section** and click on **App client settings**
+- Look for your app client named `Strapi Auth` and enable Cognito User Pool by checking it in the **Enabled Identity Providers** section of your newly created App client
+- Feel in your callback URL and Sign out URL with the value `http://localhost:1337/connect/cognito/callback` or the one provided by your AWS Cognito provider in Strapi
```
yes good catch

alexandrebodin commented:
> Hi,
> Thanks for you contribution !
> It doesn't work on my side, I've been through the whole process but at the end I have "unknown provider". I think some code is missing in the PR here: https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/services/Providers.js#L452

I tested it and it works fine.

alexandrebodin commented:
As @petersg83 mentionned you need to add a handler once you get the access_token back from aws cognito and call the callback with the right user infos :).

ralphsomeday commented:
Yes I am on it

alexandrebodin commented:
FYI, when you customize the client redirect URI you don't have that issue as you handle the access_token directly in your front-end app

ralphsomeday commented:
Yes that is what I assumed. The access_token will be sent to frontend and then they can handle the api calls from the front end. I am now getting a 431 Request Header Fields Too Large as the access token from Cognito is quite Large more than 7K characters. How would you deal with this limitation on the server side @alexandrebodin ?

alexandrebodin commented:
When are you getting this error exactly ?

ralphsomeday commented:
When passing the access token to the connect endpoint via the query parameter

alexandrebodin commented:
Hum I don't get an error. do you run this under a proxy ?

ralphsomeday commented:
Can you share you configuration for the provider?

petersg83 commented:
For information, I tested the PR with this front-end example : https://github.com/strapi/strapi-examples/tree/master/login-react. I just added `'cognito'` there: https://github.com/strapi/strapi-examples/blob/master/login-react/src/pages/Home.js#L13

ralphsomeday commented:
Thanks @petersg83 
Just updated the doc taking into account your input and added a case for the Cognito provider. Please let me know if now all is working fine. thanks.

petersg83 commented on docs/v3.x/plugins/users-permissions.md:
```diff
@@ -336,20 +336,20 @@ The use of `ngrok` is not needed.
   - **Client Secret**: fill in the **App client secret** (`19c5c78dsfsdfssfsdfhpdb4nkpb145vesdfdsfsffgh7vwd6g45jlipbpb`)
   - **Host URI (Subdomain)**: fill in the URL value that you copied earlier (`myapp67b50345-67b50b17-local.auth.eu-central-1.amazoncognito.com`)
-  - **The redirect URL to your front-end app**: `http://localhost:3000/connect/cognito/redirect`
+  - **The redirect URL to your front-end app**: `http://localhost:1337/auth/cognito/callback`
```
Why did you change? It was the correct value: `http://localhost:3000/connect/cognito/redirect`

ralphsomeday replied:
For testing on my local dev i used this redirect URL and it worked pretty well and created the user in strapi and returned the strapi jwt with the profile information. It is a backend URL I know but for testing purpose if you do not have a frontend then this one works and after you can change it to your front end.

petersg83 replied:
Thanks I understand! However, to be consistent with the other providers and to follow the standard flow can you put it back? It is supposed to work along with the strapi-example https://github.com/strapi/strapi-examples/tree/master/login-react/ :)

petersg83 commented on packages/strapi-plugin-users-permissions/services/Providers.js:
```diff
@@ -161,6 +162,23 @@ const getProfile = async (provider, query, callback) => {
     }
+    case 'cognito': {
+      // get the id_token
+      const idToken = query.id_token;
```
`id_token` doesn't seem to exist. In the query to `auth/cognito/callback` I see the params `access_token`, `refresh_token`, `raw[access_token]`, `raw[refresh_token]`, `raw[expires_in]` and `raw[token_type]`.

Also, to do like the other providers, the code here is supposed to make a call to the cognito API to retrieve the username.

Does it work on your side with this front-end app: https://github.com/strapi/strapi-examples/tree/master/login-react ?

ralphsomeday replied:
you have to empty your core_store from the grant permissions as I had to update the provider oauth scope in the bootstrap file in order to retrieve the user profile information. So instead of only email we now retrieve the openid and profile scopes and you will now receive the id_token back from cognito.

petersg83 replied:
Thank, you are right and it works great with the front-end app now !

rmaroun replied:
@petersg83 and @alexandrebodin can we merge?

petersg83 replied:
Yes, it's good for me :) Just waiting for this: https://github.com/strapi/strapi/pull/6917#discussion_r499655530

petersg83 replied:
Also after that, can you update with master, squash your commits into one and signing-it off so the DCO pass?

rangoc replied:
> Yes that is what I assumed. The access_token will be sent to frontend and then they can handle the api calls from the front end. I am now getting a 431 Request Header Fields Too Large as the access token from Cognito is quite Large more than 7K characters. How would you deal with this limitation on the server side @alexandrebodin ?

Sorry to barge in on the conversation like this, but this topic here is the only resource that's mentioning the problem occurring to me. I've tried implementing Cognito as authentication provider for Strapi, following the documentation, and using login-react as example for front-end, and I am experiencing the same error 431 Request Header Fields Too Large that you've experienced in the message above. Could you elaborate a bit more on how you solved it and what should be done to fix it? 

Steps to reproduce the error are the following: 

1. On front end, when you click on "Connect with Cognito" button you get redirected to Cognito's sign-in/sign-up form. 
2. When you choose Sign-up, create a user, and enter the verification code, apparently user gets successfully created in Cognito's User Pool ( under the hub ) but in browser, redirection happens and redirection leads to 431 Request Header Fields Too Large. 
3. When you navigate back to the URL of frontend ( ex. https://localhost:3000 ) and click again on "Connect with Cognito" you get successfully redirected to the front-end's page that's showing  successful log-in message with your Strapi's username in it.

ralphsomeday replied:
@rangoc this limitation seems to come from node and I think it was introduced in version 10. To overcome this I had to run the following command `node --max-http-header-size=16384 ./node_modules/strapi/bin/strapi develop` try it and let me know if that works for you.

rangoc replied:
@ralphsomeday  Error still persists, version of Node that I am currently using is v12.16.1.

rangoc replied:
> @ralphsomeday Error still persists, version of Node that I am currently using is v12.16.1. I think I should update my Node version to > 14 in order to be able to set --max-http-header-size=16384

Now it works 👍

rangoc replied:
@ralphsomeday Sorry for taking this conversation in here, if it's not a good place to post this please tell me and I will delete this comment. 

I just wanted to ask, would it be possible to implement this whole flow ( the one that's implemented by react-login example ) successfully while avoiding Cognito's Hosted UI, and instead using my own custom built UI ( that I've implemented by using SDK, which enables me to authenticate with Cognito without Hosted UI )? What would be the additional steps needed to be taken for Strapi's authentication part? 

Once again, sorry for opening this topic in here, sources are scarce.

ralphsomeday replied:
Good question! You will have to research this one as usually the flow allows you to convert a cognito token into a strapi token which you then use to call the strapi api endpoints. If you authenticate using amplify or the sdk you will get a cognito token that won't be recognised by strapi. Maybe after getting the cognito token, you can call the redirect url to of the cognito provider with the access_token?

rangoc replied:
Hm, sounds like an idea :) If I try and when I try I will definitely write about it. Thanks :)
